test/src/specs/tx_pool/descendant.rs

```diff
@@ -0,0 +1,261 @@
+
+        // 1. Put `tx_family` into pending-pool.
+        let family = prepare_txfamily(node);
+        node.submit_transaction(family.a());
+        node.submit_transaction(family.b());
+
+        // 2. Propose and commit `tx_family.a` only
+        node.submit_block(&propose(node, &[family.a()]));
+        (0..window.closest()).for_each(|_| {
+            node.submit_block(&blank(node));
+        }); // continuously submit blank blocks.
+        node.submit_block(&commit(node, &[family.a()]));
+
+        let block = node.new_block(None, None, None);
+        assert_eq!(
+            vec![family.b().proposal_short_id()]
+                .into_iter()
+                .collect::<HashSet<_>>(),
+            block.union_proposal_ids(),
+        );
+        node.submit_block(&block);
+    }
+}
+
+pub struct ProposeOutOfOrder;
+
+impl Spec for ProposeOutOfOrder {
+    crate::name!("propose_out_of_order");
+
+    // Case: Even if the proposals is out of order of relatives(child transaction
+    //       proposed before its parent transaction), miner commits in order of
+    //       relatives
+    //   1. Put `tx_family` into pending-pool.
+    //   2. Propose `[tx_family.b, tx_family.a]`, then continuously submit blank blocks.
+    //   3. Expect committing `[tx_family.a, tx_family.b]`.
+    fn run(&self, net: &mut Net) {
+        let node = &net.nodes[0];
+        let window = node.consensus().tx_proposal_window();
+        node.generate_blocks((window.farthest() + 2) as usize);
+
+        // 1. Put `tx_family` into pending-pool.
+        let family = prepare_txfamily(node);
+        node.submit_transaction(family.a());
+        node.submit_transaction(family.b());
+
+        // 2. Propose `[tx_family.b, tx_family.a]`, then continuously submit blank blocks.
+        node.submit_block(&propose(node, &[family.b(), family.a()]));
+        (0..window.closest()).for_each(|_| {
+            node.submit_block(&blank(node)); // continuously submit blank blocks.
+        });
+
+        // 3. Expect committing `[tx_family.a, tx_family.b]`.
+        let block = node.new_block(None, None, None);
+        assert_eq!(
+            [family.a().to_owned(), family.b().to_owned()],
+            block.transactions()[1..],
+        );
+        node.submit_block(&block);
+    }
+}
+
+pub struct SubmitTransactionWhenItsParentInGap;
+
+impl Spec for SubmitTransactionWhenItsParentInGap {
+    crate::name!("submit_transaction_when_its_parent_in_gap");
+
+    // Case: This case intends to test that submit a transaction which its parent is in gap-pool
+    fn run(&self, net: &mut Net) {
+        let node = &net.nodes[0];
+        let window = node.consensus().tx_proposal_window();
+        node.generate_blocks((window.farthest() + 2) as usize);
+
+        // 1. Propose `tx_family.a` into gap-pool.
+        let family = prepare_txfamily(node);
+        node.submit_transaction(family.a());
+        node.submit_block(&propose(node, &[family.a()]));
```

Comment:
I think after a new block, the `a` will be in the gap-pool.

Reply:
Yes. But here explicitly proposing is better.

Reply:
I meant that one more block after proposing.